pkg/registry/apis/datasource/plugincontext.go

```diff
@@ -2,10 +2,117 @@ package datasource
+	"github.com/grafana/grafana/pkg/services/grafana-apiserver/utils"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/plugincontext"
+)
+
+// This provides access to settings saved in the database.
+// Authorization checks will happen within each function, and the user in ctx will
+// limit which namespace/tenant/org we are talking to
+type PluginConfigProvider interface {
+	// Datasource gets all data source plugins (with elevated permissions).
+	GetDataSource(ctx context.Context, pluginID, uid string) (*v0alpha1.DataSourceConnection, error)
+
+	// Datasources lists all data sources (with elevated permissions).
+	ListDatasources(ctx context.Context, pluginID string) (*v0alpha1.DataSourceConnectionList, error)
+
+	// Return settings (decrypted!) for a specific plugin
+	// This will require "query" permission for the user in context
+	PluginContextForDataSource(ctx context.Context, pluginID, uid string) (backend.PluginContext, error)
+}
+
+func ProvideDefaultPluginConfigs(
+	dsService datasources.DataSourceService,
+	dsCache datasources.CacheService,
+	contextProvider *plugincontext.Provider) PluginConfigProvider {
+	return &defaultPluginConfigProvider{
+		dsService:       dsService,
+		dsCache:         dsCache,
+		contextProvider: contextProvider,
+	}
+}
+
+type defaultPluginConfigProvider struct {
+	dsService       datasources.DataSourceService
+	dsCache         datasources.CacheService
+	contextProvider *plugincontext.Provider
+}
+
+var (
+	_ PluginConfigProvider = (*defaultPluginConfigProvider)(nil)
 )
 
-type PluginContextProvider interface {
-	PluginContextForDataSource(ctx context.Context, pluginID, name string) (backend.PluginContext, error)
+func (q *defaultPluginConfigProvider) GetDataSource(ctx context.Context, pluginID, uid string) (*v0alpha1.DataSourceConnection, error) {
+	info, err := request.NamespaceInfoFrom(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+	user, err := appcontext.User(ctx)
+	if err != nil {
+		return nil, err
+	}
+	ds, err := q.dsCache.GetDatasourceByUID(ctx, uid, user, false)
+	if err != nil {
+		return nil, err
+	}
+	return asConnection(ds, info.Value)
+}
+
+func (q *defaultPluginConfigProvider) ListDatasources(ctx context.Context, pluginID string) (*v0alpha1.DataSourceConnectionList, error) {
+	info, err := request.NamespaceInfoFrom(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+
+	ds, err := q.dsService.GetDataSourcesByType(ctx, &datasources.GetDataSourcesByTypeQuery{
+		OrgID: info.OrgID,
+		Type:  pluginID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return asConnectionList(ds, info.Value)
+}
+
+func (q *defaultPluginConfigProvider) PluginContextForDataSource(ctx context.Context, pluginID, uid string) (backend.PluginContext, error) {
+	return q.contextProvider.PluginContextForDataSource(ctx, pluginID, uid)
+}
+
+func asConnection(ds *datasources.DataSource, ns string) (*v0alpha1.DataSourceConnection, error) {
+	v := &v0alpha1.DataSourceConnection{
+		ObjectMeta: metav1.ObjectMeta{
```

Comment:
@wbrowne -- this gets easier because I removed the TypeMeta.  K8s will fill it in automatically for the top level response, so I think we are safe.